cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/MetadataSearchClient.java

```diff
@@ -0,0 +1,133 @@
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+/**
+ * SearchClient for internal searching from metadata service
+ */
+public class MetadataSearchClient {
+
+  private static final String QUERY_PARAM_FORMAT = "%s=%s";
+  private static final String METADATA_SEARCH_BASE = "metadata/search";
+  public static final String PARAM_DELIMITER = "&";
+  public static final String PARAM_SEPARATOR = "?";
+  private final RemoteClient remoteClient;
+  private static final Gson GSON = new GsonBuilder().create();
+
+  @Inject
+  public MetadataSearchClient(DiscoveryServiceClient discoveryClient) {
+    this.remoteClient = new RemoteClient(discoveryClient, Constants.Service.METADATA_SERVICE,
+                                         new DefaultHttpRequestConfig(false), Constants.Gateway.API_VERSION_3);
+  }
+
+  /**
+   * Sends a HTTP GET request to metadata service with search params
+   *
+   * @param request {@link SearchRequest}
+   * @return {@link MetadataSearchResponse}
+   * @throws IOException
+   */
+  public MetadataSearchResponse search(SearchRequest request) throws IOException {
+    HttpRequest httpRequest = remoteClient
+      .requestBuilder(HttpMethod.GET, buildRequestURL(METADATA_SEARCH_BASE, request)).build();
+    HttpResponse response = remoteClient.execute(httpRequest);
+    return GSON.fromJson(response.getResponseBodyAsString(), MetadataSearchResponse.class);
+  }
+
+  private String buildRequestURL(String base, SearchRequest request) throws UnsupportedEncodingException {
+    StringBuilder builder = new StringBuilder();
+    builder.append(base);
+    builder.append(PARAM_SEPARATOR);
+    String simpleParamString = getSimpleParams(request).entrySet().stream()
+      .filter(this::isPresent)
+      .map(this::getFormattedParam)
+      .collect(Collectors.joining(PARAM_DELIMITER));
+    String multiParamsString = getMultiParams(request).entrySet().stream()
+      .filter(this::isSetPresent)
+      .map(this::getFormattedParamset)
+      .collect(Collectors.joining(PARAM_DELIMITER));
+    builder.append(simpleParamString);
+    builder.append(PARAM_DELIMITER);
+    builder.append(multiParamsString);
+    return builder.toString();
+  }
+
+  private boolean isPresent(Map.Entry<String, String> paramEntry) {
+    return paramEntry.getValue() != null && !paramEntry.getValue().isEmpty();
+  }
+
+  private boolean isSetPresent(Map.Entry<String, Set<String>> paramEntry) {
+    return paramEntry.getValue() != null && !paramEntry.getValue().isEmpty();
+  }
+
+  private String getFormattedParam(Map.Entry<String, String> paramEntry) {
+    return String.format(QUERY_PARAM_FORMAT, paramEntry.getKey(), paramEntry.getValue());
```

Comment:
Query parameter should be encoded using `URLEncoder`.

Reply:
Done